bootloader/lkenv/lkenv.go

```diff
@@ -0,0 +1,355 @@
+	Snap_try_gadget [SNAP_NAME_MAX_LEN]byte
+
+	/**
+	Reboot reason
+	Optional parameter to signal bootloader alternative reboot reasons
+	e.g. recovery/factory-reset/boot asset update
+	*/
+	Reboot_reason [SNAP_NAME_MAX_LEN]byte
+
+	/**
+	 Matrix for mapping of boot img partion to installed kernel snap revision
+	 At image build time:
+			 - snap prepare populates boot args as for other bootloaders
+			 - first column with bootimage part names can be filled at gadget builds
+			   or in the future by snap prepare image time
+			 - ExtractKernelAssets fills mapping for initial boot image
+	 snapd:
+			 - when new kernel snap is installed, ExtractKernelAssets updates mapping
+			   in matrix for bootloader to pick correct kernel snap to use for boot
+			 - snap_mode, snap_try_kernel, snap_try_core behaves same way as with u-boot
+			 - boot partition labels are never modified by snapd at run time
+	 bootloader:
+			 - Finds boot partition to use based on info in matrix and snap_kernel / snap_try_kernel
+			 - bootloaer does not alter matrix, only alters snap_mode
+
+	 [ <bootimg 1 part label> ] [ <currently installed kernel snap revison> ]
+	 [ <bootimg 2 part label> ] [ <currently installed kernel snap revision> ]
+	*/
+	Bootimg_matrix [SNAP_BOOTIMG_PART_NUM][2][SNAP_NAME_MAX_LEN]byte
+
+	/* name of the boot image from kernel snap to be used for extraction
+	   when not defined or empty, default boot.img will be used */
+	Bootimg_file_name [SNAP_NAME_MAX_LEN]byte
+
+	/**
+	GADGET assets: Matrix for mapping of gadget asset partions
+	Optional boot asset tracking, based on bootloader support
+	Some boot chains support A/B boot assets for increased robustness
+	example being A/B TrustExecutionEnvironment
+	This matrix can be used to track current and try boot assets for
+	robust updates
+
+	[ <boot assets 1 part label> ] [ <currently installed assets revison> ]
+	[ <boot assets 2 part label> ] [ <currently installed assets revision> ]
+	*/
+	Boot_asset_matrix [SNAP_BOOTIMG_PART_NUM][2][SNAP_NAME_MAX_LEN]byte
+
+	/* unused placeholders for additional parameters in the future */
+	Unused_key_1 [SNAP_NAME_MAX_LEN]byte
+	Unused_key_2 [SNAP_NAME_MAX_LEN]byte
+	Unused_key_3 [SNAP_NAME_MAX_LEN]byte
+	Unused_key_4 [SNAP_NAME_MAX_LEN]byte
+
+	/* unused array of 10 key value pairs */
+	Kye_value_pairs [10][2][SNAP_NAME_MAX_LEN]byte
+
+	/* crc32 value for structure */
+	Crc32 uint32
+}
+
+// Env contains the data of the uboot environment
+// path can be file or partition device node
+type Env struct {
+	path    string
+	pathbak string
+	env     SnapBootSelect
+}
+
+//helper function to trim string from byte array to actual length
+func cToGoString(c []byte) string {
+	if end := bytes.IndexByte(c, 0); end >= 0 {
+		return string(c[:end])
+	}
+	// XXX: no trailing \0 - should we return "" here?
+	return string(c)
+}
+
+// helper function to copy string into array and making sure it's terminated
+func copyString(b []byte, s string) {
```

Comment:
which case do you mean if string is empty, copy will not copy anything, if string is longer than byte array, it will only copy what fits in. Only case I can think of trailing zero
```

Reply:
yessish, missing test?

Reply:
there was test, but test was testing for panic :)
I updated now implementation that it concatenate string to fit into array
and added test to validate there is no panic and more test of extreme cases